yjit/src/codegen.rs

```diff
@@ -8006,6 +8006,12 @@ fn gen_send_dynamic<F: Fn(&mut Assembler) -> Opnd>(
     // Save PC and SP to prepare for dynamic dispatch
     jit_prepare_non_leaf_call(jit, asm);
 
+    // Squash stack canary that might be left over from elsewhere
+    assert_eq!(false, asm.get_leaf_ccall());
+    if cfg!(debug_assertions) {
+        asm.store(asm.ctx.sp_opnd(0), 0.into());
```

Comment:
Looking at past backtraces, fallbacks are indeed suspicious sources. But I'm not sure if this helps. If we set a canary on `asm.ccall`, it should have been cleared if `longjmp` did not happen. If `longjmp` did happen, however, it wouldn't reach this fallback code. WDYT?

Reply:
Though, if the canary is not about the one set by a ccall prior to this fallback but something set by a completely irrelevant block (that was left after a longjmp), that might make sense 🤔 So maybe this is necessary.

To give us more information, could we insert a C call here to trigger `rb_bug` right away if the canary is set, instead of suppressing the assertion in the fallback?

Reply:
The situation I'm imaging is some canary setting `asm.ccall` raises, which leaves the canary on the stack. Later, the stack grows to be just the same size again, and fallback checks the canary by accident.

Reply:
Yeah, so I think you thought the same thing as https://github.com/ruby/ruby/pull/10136#discussion_r1506710266.

Before just clearing it, could we add a runtime assertion here (using a C call that may call `rb_bug`) so that we'll know we indeed need to clear it?

Reply:
> To give us more information, could we insert a C call here to trigger rb_bug right away if the canary is set, instead of suppressing the assertion in the fallback?

I think that would give us mostly the same backtrace, because the fallback functions pretty much all start by pushing a frame, and that's when they check for the canary.

To find who left the canary I think we'd need to check in rb_longjmp()

Reply:
I agree it's the case for cases like [this](https://github.com/ruby/ruby/actions/runs/8083502532/job/22086714152) where it's coming from a fallback function `rb_vm_opt_send_without_block`. However, failures on `webauthn_poller` like [this](https://github.com/ruby/ruby/actions/runs/7992014651/job/21824580101#step:13:101) don't have a fallback function in the backtrace, so the theory doesn't necessarily apply to all cases.

But, hmm, I guess it doesn't really help both cases 🤷 So I guess it's not worth an assertion unless we do it earlier at longjmp.